rclpy/rclpy/graph_listener.py

```diff
@@ -0,0 +1,197 @@
+# Copyright 2017 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import threading
+import traceback
+
+from rclpy.impl.implementation_singleton import rclpy_implementation as _rclpy
+import rclpy.logging
+from rclpy.timer import WallTimer as _WallTimer
+from rclpy.wait_set import WaitSet as _WaitSet
+
+
+class GraphListenerSingleton:
```

Comment:
How is this singleton being destroyed in case of e.g. shutting down rclpy? When is the thread being stopped?

Reply:
In d10ee26eec13cd29bf498209f939e2bdf85092c4 it checks if rclpy was shut down once per second.